includes/embeds/class-amp-gallery-embed-handler.php

```diff
@@ -16,264 +15,174 @@
-				'include' => '',
-				'exclude' => '',
-				'size'    => [ $this->args['width'], $this->args['height'] ],
-				'link'    => 'none',
-			],
-			$attr,
-			'gallery'
-		);
+		$is_lightbox = isset( $attrs['amp-lightbox'] ) && rest_sanitize_boolean( $attrs['amp-lightbox'] );
 
-		if ( ! empty( $attr['amp-lightbox'] ) ) {
-			$atts['lightbox'] = rest_sanitize_boolean( $attr['amp-lightbox'] );
+		if ( ! $is_carousel && ! $is_lightbox ) {
+			return $html;
 		}
 
-		$id = (int) $atts['id'];
-
-		if ( ! empty( $atts['include'] ) ) {
-			$attachments = get_posts(
-				[
-					'include'        => $atts['include'],
-					'post_status'    => 'inherit',
-					'post_type'      => 'attachment',
-					'post_mime_type' => 'image',
-					'order'          => $atts['order'],
-					'orderby'        => $atts['orderby'],
-					'fields'         => 'ids',
-				]
-			);
-		} elseif ( ! empty( $atts['exclude'] ) ) {
-			$attachments = get_children(
-				[
-					'post_parent'    => $id,
-					'exclude'        => $atts['exclude'],
-					'post_status'    => 'inherit',
-					'post_type'      => 'attachment',
-					'post_mime_type' => 'image',
-					'order'          => $atts['order'],
-					'orderby'        => $atts['orderby'],
-					'fields'         => 'ids',
-				]
-			);
-		} else {
-			$attachments = get_children(
-				[
-					'post_parent'    => $id,
-					'post_status'    => 'inherit',
-					'post_type'      => 'attachment',
-					'post_mime_type' => 'image',
-					'order'          => $atts['order'],
-					'orderby'        => $atts['orderby'],
-					'fields'         => 'ids',
-				]
-			);
+		if ( $is_carousel ) {
+			$gallery_size = isset( $attrs['size'] ) ? $attrs['size'] : null;
+
+			if ( $gallery_size && 'thumbnail' === $gallery_size ) {
+				/*
+				 * If the 'gallery' shortcode has a `size` attribute of `thumbnail`, prevent outputting an <amp-carousel>.
+				 * That will often get thumbnail images around 150 x 150,
+				 * while the <amp-carousel> usually has a width of 600 and a height of 480.
+				 * That often means very low-resolution images.
+				 * So fall back to the normal 'gallery' shortcode callback, gallery_shortcode().
+				 */
+				return $html;
+			}
+
+			if ( ! $gallery_size ) {
+				// Default to `large` if no `size` attribute is specified.
+				$attrs['size'] = 'large';
+			}
 		}
 
-		if ( empty( $attachments ) ) {
-			return '';
+		if ( $is_lightbox ) {
+			// Prevent wrapping the images in anchor tags if a lightbox is specified. If that is done the link will get
+			// preference over the lightbox when the image is clicked.
+			$attrs['link'] = 'none';
 		}
 
-		$urls = [];
-		foreach ( $attachments as $attachment_id ) {
-			list( $url, $width, $height ) = wp_get_attachment_image_src( $attachment_id, $atts['size'], true );
+		// Use `data` attributes to indicate which options are configured for the embed. These indications are later
+		// processed during sanitization of the embed in `::sanitize_raw_embeds`.
+		add_filter(
```

Comment:
This filter is added but not removed. Wouldn't this cause unexpected results if multiple `gallery` shortcodes appear on the same page?

Reply:
Addressed in bab1fc9edbc751938dd9f804c62b60ab950d04d9

Reply:
Yea good catch.